Library/OcVariableRuntimeLib/OcVariableRuntimeLib.c

```diff
@@ -0,0 +1,621 @@
+               );
+  }
+
+  if (EFI_ERROR (Status)) {
+    NvramDir->Close (NvramDir);
+    OcAsciiStringBufferFree (&Context.StringBuffer);
+    return Status;
+  }
+
+  DeleteFile (NvramDir, OPEN_CORE_NVRAM_FILENAME);
+  STATIC_ASSERT (NVRAM_PLIST_MAX_SIZE <= MAX_UINT32, "NVRAM_PLIST_MAX_SIZE must be less than or equal to UINT32_MAX");
+  if (Context.StringBuffer->StringLength > NVRAM_PLIST_MAX_SIZE) {
+    Status = EFI_OUT_OF_RESOURCES;
+  } else {
+    Status = OcSetFileData (
+               NvramDir,
+               OPEN_CORE_NVRAM_FILENAME,
+               Context.StringBuffer->String,
+               (UINT32)Context.StringBuffer->StringLength
+               );
+  }
+
+  OcAsciiStringBufferFree (&Context.StringBuffer);
+  NvramDir->Close (NvramDir);
+
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+ResetNvram (
+  VOID
+  )
+{
+  EFI_STATUS         Status;
+  EFI_STATUS         AltStatus;
+  EFI_FILE_PROTOCOL  *NvramDir;
+
+  Status = LocateNvramDir (&NvramDir);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status    = DeleteFile (NvramDir, OPEN_CORE_NVRAM_FILENAME);
+  AltStatus = DeleteFile (NvramDir, OPEN_CORE_NVRAM_FALLBACK_FILENAME);
+
+  NvramDir->Close (NvramDir);
+
+  return EFI_ERROR (Status) ? Status : AltStatus;
+}
+
+//
+// If Luanchd.command is installed this should correctly handle reboots during full or partial OTA updates.
+// When installing from USB we will likely go to the wrong OS after first reboot: the one in nvram.fallback,
+// if present (which might or might not be the drive we installed to); otherwise the empty NVRAM default OS
+// (likewise). Apart from needing to manually select that, at some point, the install should otherwise be correct.
+//
+STATIC
+EFI_STATUS
+EFIAPI
+SwitchToFallback (
+  VOID
+  )
+{
+  EFI_STATUS         Status;
+  EFI_FILE_PROTOCOL  *NvramDir;
+  UINT8              *FileBuffer;
+  UINT32             FileSize;
+
+  Status = LocateNvramDir (&NvramDir);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Given that this approach is designed to avoid continually displaying 'macOS Installer' option,
+  // we want to switch back to empty NVRAM defaults even if nvram.fallback does not exist, so we
+  // do not check that.
+  //
+  FileBuffer = OcReadFileFromDirectory (NvramDir, OPEN_CORE_NVRAM_FILENAME, &FileSize, NVRAM_PLIST_MAX_SIZE);
+  if (FileBuffer == NULL) {
+    NvramDir->Close (NvramDir);
+    return EFI_ALREADY_STARTED;
+  }
+
+  DeleteFile (NvramDir, OPEN_CORE_NVRAM_USED_FILENAME);
+  DeleteFile (NvramDir, OPEN_CORE_NVRAM_FILENAME);
+  Status = OcSetFileData (
+             NvramDir,
+             OPEN_CORE_NVRAM_USED_FILENAME,
+             FileBuffer,
+             FileSize
+             );
+
+  NvramDir->Close (NvramDir);
+  FreePool (FileBuffer);
+
+  return Status;
+}
+
+STATIC
+OC_VARIABLE_RUNTIME_PROTOCOL
+  mOcVariableRuntimeProtocol = {
+  OC_VARIABLE_RUNTIME_PROTOCOL_REVISION,
+  (OC_VARIABLE_RUNTIME_PROTOCOL_LOAD_NVRAM)LoadNvram,
```

Comment:
There should not be needed a cast. Why?

Reply:
Hi. A cast is not needed. But I noticed that a few libraries in EDK-II use this convention. The point (at least as I've understood it, and why I've added it) it is to allow a search for `OC_VARIABLE_RUNTIME_PROTOCOL_LOAD_NVRAM` to find `LoadNvram` (and same for the others). Otherwise both `LoadNvram`, and its use in constructing this protocol, only _implicitly_ reference the type (i.e. no global search for the type will find it, even though its signature has to match the type, in fact).

Reply:
That „convention“ is dangerous because it may hide type mismatches, including calling convention. It’s usually best to do the opposite of EDK II. :)

Reply:
Yes, I already spotted that that was the countervailing reason _not_ to use it. It is a shame to have no way to get both advantages, but I don't think there is one. (Unless using `///<` comment???) I'll remove it.

Reply:
Unsure, I haven‘t ever felt a need for that. I‘d search for the protocol over the function type. But I don‘t mind a comment (but do mind the cast).

Reply:
I'll just remove. Obviously it's findable - as you say, by searching for the protocol - it's just an extra step or two. (Steps which I find myself doing, quite often!!) (But unless it's done consistently everywhere, having it sort of findable another way in some places is probably misleading, anyway.)

Reply:
Removed.